---
title: "Targets factories"
abstract: |
  The intended usage of scitargets is to use the targets factories such as *"tar_demultiplex_hto"* to realize the analyses. However, there are some case for which you may want to more control over the pipelines

  In this vignette, I will show minimal examples of the targets factories and equivalent code using targets or a direct call to the package functions used in the pipeline.
format:
  html:
   toc: true
   toc-depth: 3
   embed-resources: true
vignette: >
  %\VignetteIndexEntry{targets-factories}
  %\VignetteEngine{quarto::html}
  %\VignetteEncoding{UTF-8}
knitr:
  opts_chunk:
    collapse: true
    comment: '#>'
---

```{r}
#| label: setup
#| include: false
#| message: false
#| warning: false
library(scitargets)
```

# tar_demultiplex_hto

## Using the target factory (intended usage)

```{r}
#| eval: false
#| include: true
library(targets)
library(tarchetypes)
library(scitargets)

tar_option_set(
  tidy_eval = T,
  seed = 5114, # main seed (step specific seed are derived from it and step name)
  format = "qs", # more efficiant than default rds
  memory = "transient", # unload the targets after the steps are completed
  packages = c("scitargets")
)
list(
  scitargets::tar_demultiplex_hto(
    run_id = "SC282_250730",
    project_id = "PRETERRAH",
    min_genes_detected = 200,
    max_genes_detected = 4000,
    mt_percent_cutoff = 5,
    singlets_dim_to_use = 1:15,
    singlets_clusters_to_use = "clusters_0.5"
  )
)
```

## Equivalent targets pipeline without the target factory

```{r}
#| eval: false
#| include: true
library(targets)
library(tarchetypes)
library(scitargets)

tar_option_set(
  tidy_eval = T,
  seed = 5114, # main seed (step specific seed are derived from it and step name)
  format = "qs", # more efficiant than default rds
  memory = "transient", # unload the targets after the steps are completed
  packages = c("scitargets")
)
list(
  targets::tar_target(
    name = parameters_SC282_250730,
    command = list(
      run = "SC282_250730",
      project = "PRETERRAH",
      min_genes_detected = 200,
      max_genes_detected = 4000,
      mt_cutoff = 5
    ),
    description = "parameters of the run SC282_250730 from the PRETERRAH project."
  ),
  targets::tar_target(
    name = cellranger_output_SC282_250730,
    command = "./data/cellranger_output/SC282_250730/",
    format = "file",
    description = "Folder containing the genes and antibodies counts generated by cellranger from the SC282_250730 run (PRETERRAH)"
  ),
  targets::tar_target_raw(
    name = seurat_obj_SC282_250730_raw,
    command = scitargets::load_seurat_data_10X(
      path = cellranger_output_SC282_250730,
      id = "PRETERRAH"
    ),
    description = "Seurat object with the counts generated by cellranger SC282_250730 from PRETERRAH"
  ),
  targets::tar_target(
    name = seurat_obj_SC282_250730_filter,
    command = scitargets::filter_cell_and_run_reduction(
      obj = seurat_obj_SC282_250730_raw,
      min_nFeature_RNA = parameters_SC282_250730[["min_genes_detected"]],
      max_nFeature_RNA = parameters_SC282_250730[["max_genes_detected"]],
      cutoff_percent_mt = parameters_SC282_250730[["mt_cutoff"]]
    ),
    description = "seurat obj from the SC282_250730 sequencing run after filtration of low quality cells based on the number of detected genes and the percentage of mitochondrial genes ( 200 < nFeature_RNA < 4000 & percent.mt < 5)."
  ),
  targets::tar_target(
    name = seurat_obj_SC282_250730_hto_demux,
    command = scitargets::demultiplex_cell(seurat_obj_SC282_250730_filter),
    description = "SC282_250730: demultiplex HTO to identify singlets, negatives, and doublets."
  ),
  targets::tar_target(
    name = seurat_obj_SC282_250730_non_neg,
    command = scitargets::extract_non_negative_cells(obj = seurat_obj_SC282_250730_hto_demux),
    description = "SC282_250730 : remove negative cells (keep singlets and doublets) and recompute reduction."
  ),
  targets::tar_target(
    name = seurat_obj_target_SC282_250730_singlets,
    command = scitargets::extract_singlets(
      obj = seurat_obj_SC282_250730_hto_demux,
      dims_to_use = 1:15
    ),
    description = "SC282_250730: remove the doublets and recompute reduction."
  ),
  targets::tar_target_raw(
    name = markers_SC282_250730,
    command = {
      Seurat::FindAllMarkers(
        seurat_obj_target_SC282_250730_singlets,
        assay = "SCT",
        group.by = cluster_to_use,
        random.seed = targets::tar_seed_get(),
        only.pos = FALSE, min.pct = 0.25, logfc.threshold = 0.25
        )
      },
    description = "SC282_250730: Markers identified for the different cells clusters (clusters_0.5)"
    ),
  targets::tar_target_raw(
    name = seurat_obj_SC282_250730_azimuth,
    command = {
      scitargets::azimuth_annot_pbmc(
        seurat_obj_target_SC282_250730_singlets,
        "clusters_0.5"
        )
      },
    description = "SC282_250730: singlets cell with azimuth celltype annotation."
    )
)
```

## Direct use of the function

```{r}
#| eval: false
#| include: true

cellranger_output_SC282_250730 <- "./data/cellranger_output/SC282_250730/"
seurat_obj_SC282_250730_raw <- targets::load_seurat_data_10X(
  path = cellranger_output_SC282_250730,
  id = "PRETERRAH"
)
seurat_obj_SC282_250730_filter <- scitargets::filter_cell_and_run_reduction(
  obj = seurat_obj_SC282_250730_raw,
  min_nFeature_RNA = 200,
  max_nFeature_RNA = 4000,
  cutoff_percent_mt = 5
)
seurat_obj_SC282_250730_hto_demux <- scitargets::demultiplex_cell(seurat_obj_SC282_250730_filter)
seurat_obj_SC282_250730_non_neg <- scitargets::extract_non_negative_cells(obj = seurat_obj_SC282_250730_hto_demux)
seurat_obj_target_SC282_250730_singlets <- scitargets::extract_singlets(
  obj = seurat_obj_SC282_250730_hto_demux,
  dims_to_use = 1:15
)
seurat_obj_SC282_250730_azimuth <- scitargets::azimuth_annot_pbmc(
  seurat_obj_target_SC282_250730_singlets,
  "clusters_0.5"
  )
```


# Optional removing of genes

The **"singlets_feat_to_remove"** argument of *"tar_demultiplex_hto"* can be set to a character vector corresponding to the names of the genes (features) names in the RNA assay of the Seurat object. In this case, the pipeline will create another object corresponding to the singlets and without the specified genes. Both the identification of markers and celltypes annotation with azimuth will be based on this new object.

It can be used to remove genes expressed on X and Y chromosomes to remove sex related batches effect.




